crates/uv-resolver/src/lock.rs

```diff
@@ -1,27 +1,133 @@
 // Temporarily allowed because this module is still in a state of flux
 // as we build out universal locking.
-#![allow(dead_code, unreachable_code)]
+#![allow(dead_code, unreachable_code, unused_variables)]
 
+use std::collections::VecDeque;
+
+use distribution_filename::WheelFilename;
 use distribution_types::{
-    BuiltDist, DirectUrlBuiltDist, DirectUrlSourceDist, Dist, DistributionMetadata, GitSourceDist,
-    IndexUrl, Name, PathBuiltDist, PathSourceDist, RegistryBuiltDist, RegistrySourceDist,
-    ResolvedDist, ToUrlError, VersionOrUrl,
+    BuiltDist, DirectUrlBuiltDist, DirectUrlSourceDist, Dist, DistributionMetadata, FileLocation,
+    GitSourceDist, IndexUrl, Name, PathBuiltDist, PathSourceDist, RegistryBuiltDist,
+    RegistrySourceDist, Resolution, ResolvedDist, ToUrlError, VersionOrUrl,
 };
 use pep440_rs::Version;
+use pep508_rs::{MarkerEnvironment, VerbatimUrl};
+use platform_tags::{TagCompatibility, TagPriority, Tags};
 use pypi_types::HashDigest;
+use rustc_hash::FxHashMap;
 use url::Url;
+use uv_normalize::PackageName;
 
 #[derive(Clone, Debug)]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[cfg_attr(feature = "serde", serde(into = "LockWire", try_from = "LockWire"))]
 pub struct Lock {
     version: u32,
-    #[cfg_attr(feature = "serde", serde(rename = "distribution"))]
     distributions: Vec<Distribution>,
+    /// A map from distribution ID to index in `distributions`.
+    ///
+    /// This can be used to quickly lookup the full distribution for any ID
+    /// in this lock. For example, the dependencies for each distribution are
+    /// listed as distributions IDs. This map can be used to find the full
+    /// distribution for each such dependency.
+    ///
+    /// It is guaranteed that every distribution in this lock has an entry in
+    /// this map, and that every dependency for every distribution has an ID
+    /// that exists in this map. That is, there are no dependencies that don't
+    /// have a corresponding locked distribution entry in the same lock file.
+    by_id: FxHashMap<DistributionId, usize>,
 }
 
 impl Lock {
-    pub(crate) fn new(mut distributions: Vec<Distribution>) -> Result<Lock, LockError> {
-        for dist in &mut distributions {
+    pub(crate) fn new(distributions: Vec<Distribution>) -> Result<Lock, LockError> {
+        let wire = LockWire {
+            version: 1,
+            distributions,
+        };
+        Lock::try_from(wire)
+    }
+
+    pub fn to_resolution(
+        &self,
+        marker_env: &MarkerEnvironment,
+        tags: &Tags,
+        root_name: &str,
+    ) -> Resolution {
+        let root = self
+            .find_by_name(root_name)
```

Comment:
Could / should this just be a special package named `root`?

Reply:
I'm not sure that privileging the notion of "root" in the lock file makes sense, particularly given that we want to have workspace support. Cargo went in the same direction:

* https://github.com/rust-lang/cargo/issues/3003
* https://github.com/rust-lang/cargo/issues/4563
* https://github.com/rust-lang/cargo/issues/3704

Reply:
Ok agreed. This makes more sense now that I see that this isn't ever intended to be part of `pip install`, but rather, something that stems from a known project root.